File: backend2/DataAccess/Common/Cache/DaMsgControlService.cs
// *******************************************************************
// 業務名称　: 互助防疫システム
// 機能概要　: メッセージコントロールマスタ
//
// 作成日　　: 2024.07.28
// 作成者　　: 蔣
// 変更履歴　:
// *******************************************************************
namespace BCC.Affect.DataAccess
{
    public static class DaMsgControlService
    {
        private const double CacheMinutes = 3d;
        private static readonly object LockObj = new();
        private static Dictionary<string, tm_afmsgctrlDto> _dic = new();
        private static DateTime _lastTime;

        public static tm_afmsgctrlDto GetMsgDto(DaDbContext db, string ctrlmsgid)
        {
            lock (LockObj)
            {
                InitCache(db);
                return _dic.TryGetValue(ctrlmsgid, out var dto) ? dto : new tm_afmsgctrlDto();
            }
        }

        private static void InitCache(DaDbContext db)
        {
            if (DaUtil.Now.Subtract(_lastTime).TotalMinutes < CacheMinutes) return;
            _dic = db.tm_afmsgctrl.SELECT.GetDtoList().ToDictionary(x => x.ctrlmsgid, x => x);
            _lastTime = DaUtil.Now;
        }

        public static TRes GetNotOkResponse<TRes>(DaDbContext db, string ctrlmsgid, params object[] param) where TRes : DaResponseBase, new()
        {
            var res = new TRes();
            //メッセージ切替区分を取得
            var msgDto = GetMsgDto(db, ctrlmsgid);

            switch (msgDto.msgkbn)
            {
                //エラーの場合
                case EnumMsgCtrlKbn.エラー:
                    res.SetServiceError(DaMessageService.GetMsg(msgDto.errormsgid, param));
                    break;
                //アラートの場合
                case EnumMsgCtrlKbn.アラート:
                    res.SetServiceAlert(DaMessageService.GetMsg(msgDto.alertmsgid, param));
                    break;
                //非表示の場合
                case EnumMsgCtrlKbn.非表示:
                default:
                    res.SetServiceHidden();
                    break;
            }
            return res;
        }
    }
}

File: backend/BusinessService/Db/Common/Cache/DaMsgControlService.vb
' *******************************************************************
' 業務名称　: 互助防疫システム
' 機能概要　: メッセージコントロールマスタ
'
' 作成日　　: 2024.07.28
' 作成者　　: 
' 変更履歴　:
' *******************************************************************
Namespace JBD.GJS.Db
    Public Module DaMsgControlService
        Private Const CacheMinutes As Double = 3R
                ''' Cannot convert FieldDeclarationSyntax, System.InvalidCastException: 型 'Microsoft.CodeAnalysis.VisualBasic.Syntax.EmptyStatementSyntax' のオブジェクトを型 'Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax' にキャストできません。
'''    場所 ICSharpCode.CodeConverter.VB.CommonConversions.RemodelVariableDeclaration(VariableDeclarationSyntax declaration)
'''    場所 ICSharpCode.CodeConverter.VB.NodesVisitor.VisitFieldDeclaration(FieldDeclarationSyntax node)
'''    場所 Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor`1.Visit(SyntaxNode node)
'''    場所 ICSharpCode.CodeConverter.VB.CommentConvertingVisitorWrapper`1.Accept(SyntaxNode csNode, Boolean addSourceMapping)
''' 
''' Input:
'''         private static readonly object LockObj = new();
''' 
''' 
                ''' Cannot convert FieldDeclarationSyntax, System.InvalidCastException: 型 'Microsoft.CodeAnalysis.VisualBasic.Syntax.EmptyStatementSyntax' のオブジェクトを型 'Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax' にキャストできません。
'''    場所 ICSharpCode.CodeConverter.VB.CommonConversions.RemodelVariableDeclaration(VariableDeclarationSyntax declaration)
'''    場所 ICSharpCode.CodeConverter.VB.NodesVisitor.VisitFieldDeclaration(FieldDeclarationSyntax node)
'''    場所 Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor`1.Visit(SyntaxNode node)
'''    場所 ICSharpCode.CodeConverter.VB.CommentConvertingVisitorWrapper`1.Accept(SyntaxNode csNode, Boolean addSourceMapping)
''' 
''' Input:
'''         private static System.Collections.Generic.Dictionary<string, JBD.GJS.Db.tm_afmsgctrlDto> _dic = new();
''' 
''' 
        Private _lastTime As Date

        Public Function GetMsgDto(db As DaDbContext, ctrlmsgid As String) As tm_afmsgctrlDto
            Dim dto As tm_afmsgctrlDto = Nothing
            'SyncLock DaMsgControlService.LockObj
            '    InitCache(db)
            '    Return If(DaMsgControlService._dic.TryGetValue(ctrlmsgid, dto), dto, New tm_afmsgctrlDto())
            'End SyncLock
        End Function

        Private Sub InitCache(db As DaDbContext)
            If DaUtil.Now.Subtract(_lastTime).TotalMinutes < CacheMinutes Then Return
            '_dic = db.tm_afmsgctrl.SELECT.GetDtoList().ToDictionary(x => x.ctrlmsgid, x => x);
            _lastTime = DaUtil.Now
        End Sub

        Public Function GetNotOkResponse(Of TRes As {DaResponseBase, New})(db As DaDbContext, ctrlmsgid As String, ParamArray param As Object()) As TRes
            Dim res = New TRes()
            'メッセージ切替区分を取得
            Dim msgDto = GetMsgDto(db, ctrlmsgid)

            Select Case msgDto.msgkbn
                'エラーの場合
                Case EnumMsgCtrlKbn.エラー
                    res.SetServiceError(DaMessageService.GetMsg(msgDto.errormsgid, param))
                'アラートの場合
                Case EnumMsgCtrlKbn.アラート
                    '非表示の場合
                    res.SetServiceAlert(DaMessageService.GetMsg(msgDto.alertmsgid, param))
                Case Else
                    res.SetServiceHidden()
            End Select
            Return res
        End Function
    End Module
End Namespace
